doc/03_research/03_planning/06_state_machine_design.md

```diff
@@ -0,0 +1,278 @@
+
+## Driving state machine
+
+![img.png](../../00_assets/Driving_SM.png)
+
+Transition:
+
+* From `Intersection state machine`
+* From `Lane change state machine`
+
+This state machine controls the speed of the ego-vehicle. It either tells the acting part of the ego vehicle to `ACCELERATE`, `BRAKE` or `KEEP` the velocity.
+
+This is assumably the main state of the state machine as the ego vehicle will most of the time follow the road.
+
+### KEEP
+
+Transition:
+
+* From `ACCEL`
+* From `BRAKE`
+
+Keep the current target speed, applied most of the time. From here changes to the `ACCEL` or `BRAKE` state are performed, if events require a change of `target_speed`.
+
+### ACCEL
+
+Transition:
+
+* From `KEEP` if `new target_speed` is greater than current `target_speed`.
+* From `BRAKE` if `new target_speed` is greater than current `target_speed`.
+
+Set a new target speed and change back to `KEEP` state afterwards.
+
+### Brake
+
+Transition:
+
+* From `KEEP` if `new target_speed` is smaller than current `target_speed` or an `obstacle` or the `leading_vehicle` is in braking distance.
+* From `BRAKE` if `new target_speed` is smaller than current `target_speed` or an `obstacle` or the `leading_vehicle` is in braking distance.
+
+Set a new target speed and change back to `KEEP` state afterwards.
+
+## Lane change state machine
+
+![img.png](../../00_assets/Lane_Change_SM.png)
+
+Transition:
+
+* From `driving state machine` by `lane_change_requested`
+
+This state machine completes the change of a lane. This is triggered from the super state machine and can have multiple triggers. Those include:
+
+* Join highway
+* Leave highway
+* RoadOption:
+  * CHANGELANELEFT
+  * CHANGELANERIGHT
+  * KEEPLANE
+* avoid obstacle(doors, static objects)
+* give way to emergency vehicle
+* overtake slow moving vehicle
+* leave a parking bay
+
+### DECIDE_LANE_CHANGE
+
+Transition:
+
+* From super state machine by above triggers
+
+From the super state machine the transition to change the lane is given by one of the above triggers. This state decides to which lane should be changed dependent on the trigger.
+It takes into account if there are lanes to the left and/or right and if the lane change is requested by a roadOption command.
+
+### CHANGE_LANE_LEFT
+
+Transition:
+
+* From `DECIDE_LANE_CHANGE` by `RoadOption.CHANGELANELEFT` or `obstacle_in_lane` or `leader_vehicle_speed < LEADERTHRESHOLD`
+
+This state performs a lane change to the lane on the left.
+
+If the lane is for oncoming traffic, it should be checked if there is enough time to complete two lane changes plus a buffer for the length of an obstacle without an accident.
+
+If the lane is for traffic flowing in the same direction, It should be checked if the lane is free and if there is no faster moving vehicles coming from behind.
+
+If an obstacle or a slow leading vehicle are the reasons for the lane change, to join the same lane after the obstacle is passed, the state changes to `CHANGE_LANE_RIGHT`.
```

Comment:
In der Grafik sollte noch in irgendeiner Form angezeigt werden, dass das erst ausgelöst wird, nachdem der Spurwechsel nach links schon vollzogen wurde. Das ist im Moment noch etwas verwirrend

Reply:
resolved by https://github.com/ll7/paf22/pull/94/commits/6e4599cf72c6736639f46ca826be2b8f68ab3a33